### ISSUE TITLE
[Feature] [Enhancement] : Input widget v2 with improved validations

### ISSUE THREAD

**Issue by sbalaji1192** [appsmithorg/appsmith#9119]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

Current input widget has lot of open issues related to validation. Fixing them will break the existing application. So let's introduce a version 2 of  input widget with improved validations.

Existing issues with input widget validation.
- #7697
- #7899
- #8012
- #7675 
- #8591
- https://github.com/appsmithorg/appsmith/issues/7475
- https://github.com/appsmithorg/appsmith/issues/9580
- #9822
- https://github.com/appsmithorg/appsmith/issues/9580

Goal of this issue is to fix the existing issues.

### Why should this be worked on?

Input widget is one of the most used widget. Improved input widget will definitely help user in creating solid applications.

**Comment by sbalaji1192** [comment]
@vivekverma2312 Could you test this on DP - https://deploy-preview-9582--appsmith-editor-ce.netlify.app/applications?

**Comment by vivekverma2312** [comment]
@sbalaji1192 Few issues observed during the DP testing
1. Auto suggestion is not working when I am trying to bind input widget in text widget
2. In case of numbers type when I delete the data from field then `0` populates and I am not able to delete that. https://www.loom.com/share/ab16dfe075dc497f8adbac48e30be895
3. Suggestion: Instead of writing {{typeof Input1.text}} can we make it {{Input1.text.type}}
4. Allow currency change is disabled by default and Allow country code change enabled by default, can we make it consistent
5. Reset on submit is not functional for currency input

Testing is still in progress

**Comment by aswathkk** [comment]
@sbalaji1192 Can we include #9822 in this?

**Comment by sbalaji1192** [comment]
@aswathkk will try to squeeze this in.

**Comment by vivekverma2312** [comment]
@sbalaji1192 #7899, #7810, #8012 still exist. In #7675 now NaN is observed on deleting the value https://www.loom.com/share/048d14e096844a2c8d41c505d770e411

**Comment by sbalaji1192** [comment]
@vivekverma2312 Could you tell me Hhw did you test #7899 and #8012 ?

**Comment by sbalaji1192** [comment]
@vivekverma2312 please test this on - https://appsmith-git-fix-improved-input-widget-validation-get-appsmith.vercel.app/applications/61976b5f86af5222e17018e0/pages/61976b5f86af5222e17018e2/edit

**Comment by vivekverma2312** [comment]
@sbalaji1192 Observed new issue after fix for issue2 
6. Default data populated after deleting the value in Input widget https://www.loom.com/share/2119a982b083410481ce0bd6be0d7e68 
#7810 and #8012 still exist

**Comment by sbalaji1192** [comment]
@vivekverma2312 fixed #6.
Regarding #7810, It seems like an issue with evaluations not with input widget itself. Will further discuss about this with @somangshu .
Regarding #8012, It is working as expected. The app provided by the user has work around in validations logic, so it is failing with input widget v2. you need to edit the logics and use appropriate dataype to make it work.

**Comment by somangshu** [comment]
@sbalaji1192 regarding #8012, does it mean we need to think about any kind of a migration. It would be unacceptable for the users to have there app broken without knowledge, Usually we solve for this. Help me understand more here.

**Comment by sbalaji1192** [comment]
@somangshu So the user who created 8012 has created an app with input widget v1 to demonstrate the issues with validations. where he wrote conditions with all input types values as datatype string, which is not the case with input widget v2. based on the input chosen, the datatype of values changes. if we use the same conditions from that app. it will fail because of mismatching datatype. we don't need any migrations for this change.

**Comment by somangshu** [comment]
I understood, The condition any ways need to be re-written on the input v2, we are good to go then

**Comment by vivekverma2312** [comment]
@sbalaji1192 Observed one crash, If we drag and drop currency input on new page then widget crashes
https://www.loom.com/share/81005c7015714602a06593aa326d76d5

**Comment by YogeshJayaseelan** [comment]
@sbalaji1192 @vivekverma2312 Selected country code/flag is not displayed in the widget and it is observed that only after refreshing the page populates the selected country code/Flag in the widget

https://loom.com/share/918996f3a9bb482487d68b3077ce0b92

**Comment by YogeshJayaseelan** [comment]
@sbalaji1192 The watermark text of tool tip field in both phone and currency widget reads about 'Password' - Can we change this with relevant watermark text

<img width="1680" alt="Screenshot 2021-12-31 at 11 23 00 AM" src="https://user-images.githubusercontent.com/35128442/147806468-9873bbd2-f868-4ccf-9c44-d0122b2d8de5.png">

**Comment by sbalaji1192** [comment]
@YogeshJayaseelan @vivekverma2312 Fixed your callouts. please test again.

**Comment by vivekverma2312** [comment]
@sbalaji1192 When I am trying to bind the Input widget to text widget I am getting auto suggestion for country code and currency country code
<img width="303" alt="Screenshot 2022-01-05 at 10 55 09 AM" src="https://user-images.githubusercontent.com/83569920/148166298-98738986-fe3c-4892-bb1e-9655be1541ab.png">

**Comment by sbalaji1192** [comment]
@vivekverma2312 Fixed. please test again.

**Comment by vivekverma2312** [comment]
@sbalaji1192 1. County code is not updating for United States for Phone Input widget
https://www.loom.com/share/45d072cd0bba42f4895e313b89d9fbfe
2. Country flag is not coming on country code drop-down options for Phone Input widget

**Comment by sbalaji1192** [comment]
@vivekverma2312 Fixed your callouts. Also fixed callouts posted by @dilippitchika over [here](https://github.com/appsmithorg/appsmith/pull/9582#issuecomment-1005686279). Please verify both.

**Comment by YogeshJayaseelan** [comment]
@sbalaji1192 Tested this PR and please find the status below,

Issue raised by @vivekverma2312 are fixed and with respect to issues raised by @dilippitchika there is one bug reopen with the below steps 

Currency widget - Add 2 decimals to default text keeping the decimal point option as 0 and check for the error message and now change the decimal point option 2 and see the error message will still displayed 

https://www.loom.com/share/a94d782baae2452aa44c72d3610fe8b0

**Comment by sbalaji1192** [comment]
@YogeshJayaseelan It's an issue on the property pane, not with the widget itself. if the dropdown value is 0, the property pane is not updating the value. Could you create a separate issue for this ?

**Comment by shwetha-ramesh** [comment]
@sbalaji1192 tested this and found below issues. 

1. When PhoneInput and Currencyinput widget are resized the border moves inside the widget. This does not happen for Input widget
![image](https://user-images.githubusercontent.com/92293815/148934965-fac88564-a6e4-4f47-aaa8-8b7e0af68d30.png)
https://loom.com/share/347a1c72c1b84be5a6b9bfa72899621b

2. When PhoneInput and Currencyinput widget are resized user is able to place Input widget on top of it.
![image](https://user-images.githubusercontent.com/92293815/148935134-8ef76cb9-d0f9-4554-bcd2-6a0d4ed4fa85.png)
https://loom.com/share/4a88834d835d416888bf9e0d1365d16d

3. There is no icon present in Entity explorer for input widget
![image](https://user-images.githubusercontent.com/92293815/148935213-1bddd35b-6ea9-4835-8c43-dd39fe6910f0.png)

4. Icons disappear when Icon alignment is set to right for Number and Password ```Data type``` in Input widget 
https://loom.com/share/4ccaeb50120248ec947e93d7d19b4db2

5. Tooltip watermark for any ```Data type``` selected in Input widget is not relevant.
![image](https://user-images.githubusercontent.com/92293815/148935503-8f9b07a2-a804-45f9-894a-edcd1772ed26.png)

6. Background color for Phoneinput and Currencyinput varies when``` Allow country code change``` and ```Allow currency change``` are enabled.
![image](https://user-images.githubusercontent.com/92293815/148935639-1718056c-010c-4116-b0cc-9e608f78e36c.png)

7. Property pane in Input widget shows same Error message watermark for all Data types.
![image](https://user-images.githubusercontent.com/92293815/148935713-90fd54b3-ad4d-456e-bfbc-8bd574074e04.png)

8. No error message when negative value is entered in ```Max Chars``` for Input widget.
![image](https://user-images.githubusercontent.com/92293815/148935774-b926cde5-ea36-42e3-8f0b-0d50f8655ec7.png)

**Comment by sbalaji1192** [comment]
@shwetha-ramesh 
WRT #1 and #2, We have country dropdown and currency dropdown associated with the widget so we can't reduce the width beyond the length of the dropdown. And Also it seems like an invalid scenario.

#4, Password show/hide icon and increment/decrement controls are occupying the right-hand side, so we can't place the icon over there - @somangshu need your input on this.

#5, We cannot change the placeholder text based on the input data type. so using a common sentence that applies to all types.

Fixed rest. please test.

**Comment by somangshu** [comment]
@sbalaji1192 here is my thought, in case of the number input type we can add a margin right to the icon such that it is not overlapped, that is in case this is the problem.

**Comment by YogeshJayaseelan** [comment]
@sbalaji1192 verified the fixed issues and they are working as expected

**Comment by btsgh** [comment]
All of the existing issues in the bug description, which this bug is supposed to address, have been verified and closed. Hence marking this as Verified as well.
